pip/pip-341.md

```diff
@@ -0,0 +1,305 @@
+
+### LoggingMetricsTrackerFactory
+The default configured metrics tracker factory will be the `LoggingMetricsTrackerFactory` as it will replace the existing functionality.
+
+The logging producer/consumer metrics tracker will initialize and hold an instance of the underlying stats recorder, and forward all calls to the stats recorder.
+
+Example:
+```java
+public class LoggingConsumerMetricsTracker implements ConsumerMetricsTracker {
+    private final ConsumerStatsRecorder statsRecorder;
+
+    public LoggingConsumerMetricsTracker(PulsarClient pulsarClient, Consumer<?> consumer) {
+        if (consumer instanceof MultiTopicsConsumerImpl) {
+            this.statsRecorder = new MultiTopicConsumerStatsRecorderImpl(consumer);
+        } else if (consumer instanceof ConsumerImpl){
+            this.statsRecorder = new ConsumerStatsRecorderImpl((PulsarClientImpl) pulsarClient, null, consumer);
+        } else {
+            this.statsRecorder = new ConsumerStatsDisabled();
+        }
+    }
+
+    @Override
+    public void recordMessageReceived(Message<?> message) {
+        statsRecorder.updateNumMsgsReceived(message);
+    }
+
+    // Other methods...
+}
+```
+
+This allows us to completely replace all existing calls to the stat recorders with our metrics trackers, while still maintaining backwards compatibility with the previous behavior.
+
+### Backwards compatibility with statsIntervalSeconds configuration
+We can mark the `statsIntervalSeconds` client configuration as deprecated since implementation-specific configurations should be provided via the params map when constructing the metrics tracker in `ClientBuilder` (`statsIntervalSeconds` is specific to `LoggingMetricsTrackerFactory`).
+
+We will set the default metrics tracker factory as the `LoggingMetricsTrackerFactory` to maintain backwards compatibility, however we still need to do some checks:
+1. If a non-default non-null metrics tracker factory (ex. OTel one) is configured, then use it
+2. Else if `statsIntervalSeconds=0`, then the user has explicitly disabled client metrics so we must use the `NoOpMetricsTrackerFactory`
+3. Else, use `LoggingMetricsTrackerFactory`
+
+Inside `LoggingMetricsTrackerFactory`, we must also derive `statsIntervalSeconds`:
+1. First try to use `statsIntervalSeconds` if it is configured via the new params map in `ClientBuilder metricsTrackerFactory(String metricsTrackerFactoryClassName, Map<String, String> params)`
+2. Otherwise, try to use the deprecated `statsIntervalSeconds` value if it is greater than 0
+3. Finally, default to 60 seconds if `statsIntervalSeconds` is not configured anywhere
+
+### Supporting existing Consumer/Producer getStats API
+There are two major challenges with supporting the existing producer/consumer `getStats()` API:
```

Comment:
If you have the two mechanisms in place, this problem disappears. This implementation will be deleted in two versions from now.